tailnet/service.go

```diff
@@ -230,3 +238,85 @@ func (c communicator) loopResp() {
 		}
 	}
 }
+
+type NetworkTelemetryBatcher struct {
+	clock     clock.Clock
+	frequency time.Duration
+	maxSize   int
+	batchFn   func(batch []*proto.TelemetryEvent)
+
+	mu      sync.Mutex
+	closed  chan struct{}
+	done    chan struct{}
+	ticker  *clock.Ticker
+	pending []*proto.TelemetryEvent
+}
+
+func NewNetworkTelemetryBatcher(clk clock.Clock, frequency time.Duration, maxSize int, batchFn func(batch []*proto.TelemetryEvent)) *NetworkTelemetryBatcher {
+	b := &NetworkTelemetryBatcher{
+		clock:     clk,
+		frequency: frequency,
+		maxSize:   maxSize,
+		batchFn:   batchFn,
+		closed:    make(chan struct{}),
+		done:      make(chan struct{}),
+	}
+	b.start()
+	return b
+}
+
+func (b *NetworkTelemetryBatcher) Close() error {
+	close(b.closed)
+	<-b.done
+	return nil
+}
```

Comment:
This should probably accept a context so it can't block forever, or move the waiting to a separate `WaitUntilEmpty()` or similar

Reply:
I'm not going to make it accept a context, but I've added a 5s timeout and an error.